src/Extensions/Compatability/Validation/Code.cs

```diff
@@ -0,0 +1,515 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.Omex.Extensions.Compatability.Validation
+{
+	/// <summary>
+	/// Code validation
+	/// </summary>
+	public static class Code
+	{
+		/// <summary>
+		/// Check object's state, throws specified Exception type if state evaluates to false.
+		/// </summary>
+		/// <typeparam name="TException">Exception to throw</typeparam>
+		/// <param name="state">Object's state</param>
+		/// <param name="errorMessage">Error message</param>
+		/// <param name="tagId">Tag Id to log, leave null if no logging is needed</param>
+		[Obsolete(ObsoleteMessage, IsObsoleteError)]
+		public static void Expects<TException>(bool state, string errorMessage, EventId? tagId)
+			where TException : Exception
+		{
+			if (!Validate(state, errorMessage, tagId))
+			{
+				ReportError<TException>(errorMessage);
+			}
+		}
+
+
+		/// <summary>
+		/// Checks that the enumerable argument is not null and doesn't contain any nulls
+		/// </summary>
+		/// <remarks>Be careful to not pass enumerables that can be enumerated only once</remarks>
+		/// <typeparam name="TEnumerable">Type of the enumerable</typeparam>
+		/// <param name="argumentValue">The argument value.</param>
+		/// <param name="argumentName">Name of the argument.</param>
+		/// <param name="tagId">Tag Id to log, leave null if no logging is needed</param>
+		/// <exception cref="ArgumentException">Thrown if any argument  <paramref name="argumentValue"/> element is null.</exception>
+		/// <exception cref="ArgumentNullException">Thrown if the supplied argument <paramref name="argumentValue"/> is null.</exception>
+		[Obsolete(ObsoleteMessage, IsObsoleteError)]
+		public static IEnumerable<TEnumerable> ExpectsAllNotNull<TEnumerable>([ValidatedNotNull] IEnumerable<TEnumerable>? argumentValue, string argumentName, EventId? tagId)
+			where TEnumerable : class
+		{
+			argumentValue = ExpectsArgumentNotNull(argumentValue, argumentName, tagId);
+
+			if (!ValidateAllNotNull(argumentValue, argumentName, tagId))
+			{
+				ReportArgumentError(argumentName, AllErrorMessage);
+			}
+
+			return argumentValue;
+		}
+
+
+		/// <summary>
+		/// Checks the collection and throws an exception if it is null, contains no values, or contains any null values
+		/// </summary>
+		/// <typeparam name="T">Type of the collection</typeparam>
```

Comment:
C# guidelines state to use a typeparam with a more descriptive name that `T`, e.g. `TArgument` or something similar.

Also, the comment is slightly off as it isn't the type of the collection but the type of the elements within the collection. For that reason, `TEnumerable` above may not be the best name either.

Applies throughout the file

Reply:
I strongly dislike all of this file and think it should be done in a different way, it's copied here for backward compatibility and I'm worried that introducing any changes into it will increase the risk of changing behavior beside that I would like to remove it as soon as possible and don't put a lot of effort into it maintenance, that's also the reason why unit test coverage is lacking for this class.
If we end up using it for a long time I would have to create a separate PR updating it and creating tests.